### PR TITLE
 Add `enable-tensorflow-gpu` build-script flag.

### PR DESCRIPTION
**Pull request by dan-zheng**
A follow-up to #17230.
The `enable-tensorflow-gpu` flag simply does the following:
* Sets the `TF_NEED_CUDA` environment variable to 1
* Adds `--config=cuda` to the TensorFlow Bazel build options
* Warns if environment variables for TensorFlow GPU support are not set
  * TF_CUDA_VERSION, TF_CUDNN_VERSION, CUDA_TOOLKIT_PATH, CUDNN_INSTALL_PATH
  * Warnings (and not errors) are produced because the `configure` script
    provides default values for all of these variables. Users may
    intentionally leave variables unspecified to use the default values.

For now, the actual configuration must be done using environment variables.
This is the only way to configure TensorFlow without requiring user input
during the build process (by running the `configure` script).

An alternative is to create individual build flags for each of the TensorFlow
environment variables (e.g. `--tf-cuda-version`).
This may be pursued later, open questions are:
* Which flags to create (the `configure` script checks for ~10 GPU related
  flags in total)
* How to handle environment variables (what if `--tf-cuda-version` is
  unspecified but `$TF_CUDA_VERSION` is?)
  * A sane option is to ignore environment variables and use just build flags
    for configuration

TensorFlow configure script: https://github.com/tensorflow/tensorflow/blob/master/configure.py

### CODE REVIEW

**Comment by dan-zheng** [comment]
When I run `swift/utils/build-script --preset tensorflow_test,gpu` on my machine, I get some nondeterministic SourceKit test failures.

@mhong, could you perhaps try building with the same preset on your GPU machine?
```
git clone https://github.com/dan-zheng/swift.git -b enable-tensorflow-gpu
./swift/utils/update-checkout --clone --scheme tensorflow --skip-repository swift
swift/utils/build-script --preset tensorflow_test,gpu
```

File: utils/build-presets.ini [hunk 1]
```diff
@@ -1804,7 +1804,7 @@ enable-tensorflow
 
 [preset: tensorflow_linux,no_test]
 mixin-preset=buildbot_linux,no_test
-enable-tensorflow
+enable-tensorflow-gpu
```
Comment (marcrasi) [inline utils/build-presets.ini]: Is this change intentional?

**Review by marcrasi — CHANGES_REQUESTED** [review]

File: utils/build-presets.ini (hunk 1 above)
Comment (dan-zheng) [reply utils/build-presets.ini]: Oops, thanks for the catch. Fixed in https://github.com/apple/swift/pull/17839/commits/9d8a1312e2e3027e17a12bf2d796a29e4f05ebb0.

**Review by marcrasi — APPROVED** [review]

File: utils/build-script-impl [hunk 2]
```diff
@@ -2329,7 +2330,15 @@ for host in "${ALL_HOSTS[@]}"; do
                         -DSWIFT_TENSORFLOW_TARGET_INCLUDE_DIR:PATH="${TENSORFLOW_TARGET_INCLUDE_DIR}"
                     )
 
-                    # Handle TensorFlow Swfit bindings file.
+                    # Propagate TensorFlow GPU support flag to CMake.
+                    if [[ "${ENABLE_TENSORFLOW_GPU}" ]] ; then
+                        cmake_options=(
+                            "${cmake_options[@]}"
+                            -DSWIFT_ENABLE_TENSORFLOW_GPU:BOOL=TRUE
```
Comment (mhong) [inline utils/build-script-impl]: why do we need this cmake option?

File: utils/build-script-impl [hunk 3]
```diff
@@ -2329,7 +2330,15 @@ for host in "${ALL_HOSTS[@]}"; do
                         -DSWIFT_TENSORFLOW_TARGET_INCLUDE_DIR:PATH="${TENSORFLOW_TARGET_INCLUDE_DIR}"
                     )
 
-                    # Handle TensorFlow Swfit bindings file.
+                    # Propagate TensorFlow GPU support flag to CMake.
+                    if [[ "${ENABLE_TENSORFLOW_GPU}" ]] ; then
```
Comment (mhong) [inline utils/build-script-impl]: who sets ${ENABLE_TENSORFLOW_GPU}?

**Comment by mhong** [comment]
@dan-zheng I'm building this preset and will update this thread.

From reading this PR, I don't see how SourceKit can be affected by this. The new preset `tensorflow_test,gpu` does 3 things IIUC:
1. set some env vars
2. set tensorflow_bazel_options with config cuda
3. code related to `${ENABLE_TENSORFLOW_GPU}` and `-DDSWIFT_ENABLE_TENSORFLOW_GPU`

I don't understand item `#3` above, and left some questions there. The other two items should not affect any builds or tests of Swift, outside of the TF repo?

File: utils/build-script-impl (hunk 2 above)
Comment (dan-zheng) [reply utils/build-script-impl]: It's not currently used, but it could be useful in CMake for gating behavior.

I originally added it so that `-DCPU` and `-DCUDA` would both be enabled when `SWIFT_ENABLE_TENSORFLOW_GPU` is true.

File: utils/build-script-impl (hunk 3 above)
Comment (dan-zheng) [reply utils/build-script-impl]: `build-script-impl` does this (capitalizes build flag names and sets them as environmental variables).

Comment (dan-zheng) [reply utils/build-script-impl]: build-script-impl does this (capitalizes build flag names and sets them as environment variables).

**Comment by dan-zheng** [comment]
@mhong This PR shouldn't cause SourceKit test failures. I just happen to observe them on my GPU VM instance, which is the one @m3bm3b set up for GPU testing.

After this PR is merged, we may need to debug those failures.

File: utils/build-script-impl (hunk 2 above)
Comment (mhong) [reply utils/build-script-impl]: SG. would you like to add a comment that this is not currently used?

File: utils/build-script-impl (hunk 3 above)
Comment (mhong) [reply utils/build-script-impl]: I see. Thanks.

**Comment by mhong** [comment]
My local tests with the preset passed:

```
lit.py: /usr/local/google/home/hongm/ssd_part/git/swift-gpu/swift/test/lit.cfg:1196: note: Available features: CMAKE_GENERATOR=Ninja, CODEGENERATOR=AArch64, CODEGENERATOR=ARM, CODEGENERATOR=Mips, CODEGENERATOR=PowerPC, CODEGENERATOR=SystemZ, CODEGENERATOR=X86, CPU=x86_64, LinuxDistribution=Debian-rodete, OS=linux-gnu, PTRSIZE=64, SWIFT_VERSION=3, asan_runtime, asserts, autolink-extract, crash-recovery, executable_test, fuzzer_runtime, no_asan, no_lto, nonexecutable_test, optimized_stdlib, optimized_stdlib_x86_64, sourcekit, swift-remoteast-test, swift_interpreter, swift_stdlib_no_asserts, swift_test_mode_optimize_none, swift_test_mode_optimize_none_x86_64, tools-release, tsan_runtime, ubsan_runtime
-- Testing: 11208 tests, 12 threads --
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
Testing Time: 521.09s
  Expected Passes    : 9941
  Expected Failures  : 64
  Unsupported Tests  : 1203
-- check-swift-validation-linux-x86_64 finished --
--- check-swift-validation-optimize-linux-x86_64 ---
```

```
lit.py: /usr/local/google/home/hongm/ssd_part/git/swift-gpu/swift/test/lit.cfg:1196: note: Available features: CMAKE_GENERATOR=Ninja, CODEGENERATOR=AArch64, CODEGENERATOR=ARM, CODEGENERATOR=Mips, CODEGENERATOR=PowerPC, CODEGENERATOR=SystemZ, CODEGENERATOR=X86, CPU=x86_64, LinuxDistribution=Debian-rodete, OS=linux-gnu, PTRSIZE=64, SWIFT_VERSION=3, asan_runtime, asserts, autolink-extract, crash-recovery, executable_test, fuzzer_runtime, no_asan, no_lto, optimized_stdlib, optimized_stdlib_x86_64, sourcekit, swift-remoteast-test, swift_interpreter, swift_stdlib_no_asserts, swift_test_mode_optimize, swift_test_mode_optimize_x86_64, tools-release, tsan_runtime, ubsan_runtime
-- Testing: 11208 tests, 12 threads --
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
Testing Time: 658.12s
  Expected Passes    : 1082
  Expected Failures  : 17
  Unsupported Tests  : 10109
-- check-swift-validation-optimize-linux-x86_64 finished --
--- Finished tests for swift ---

**Comment by mhong** [comment]
Great work!

**Review by mhong — APPROVED** [review]

File: utils/build-script-impl (hunk 2 above)
Comment (dan-zheng) [reply utils/build-script-impl]: Done in https://github.com/apple/swift/pull/17839/commits/71386d2bd4e23db27357142389d120d67fb7a5b4.

**Comment by dan-zheng** [comment]
Thanks @mhong for reviewing and testing!
